docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+# InstructLab SDK
+
+## Motivation
+
+Today, the only way to "drive" the InstructLab opinionated workflow is via the `ilab` CLI. While this process provides a succinct way for everyday users to initialize a config, generate synthetic data, train a model, and evaluate it: the guardrails are quite limiting both in what a user can do and what the development team can add as features exposed directly to the user over time.
+
+Additionally, current consumers of InstructLab are finding themselves importing different library private and public APIs in combination with CLI functionality to achieve the workflows they want. While these more advanced usage patterns are not for everyone, providing ways to run bespoke and piecemeal workflows in a standardized and safe way for the community is a necessity.
+
+Unifying these various ranges of advanced workflows under an overarching `InstructLab SDK API` will allow for new usage patterns and a clearer story on what InstructLab can and should provide as user accessible endpoints.
+
+While each library can and _should_ have their own publicly accessible API, not all functionality being added to SDG, Training, and Eval needs to be correlated directly to the "InstructLab workflow". This Python SDK should, as the CLI does, expose an opinionated flow that uses functionality from the various libraries. The InstructLab SDK should be derived from the library APIs, not the other way around. SDG, for example, currently has a `generate_data` method, meant to only be accessed by InstructLab. This method simply calls other publicly available SDG functionality. Orchestration of the InstructLab flow like this should not be of concern to the individual libraries and instead be handled by the overarching InstructLab SDK which will maintain the user contracts. The InstructLab SDK will need to work within the bounds of what the Libraries expose as public APIs.
+
+The benefit of the above is that the opinionated flow can be accessed in a more nuanced and piecemeal way while also gaining the potential for more advanced features. Say a consumer wants to:
+
+1. Setup a custom config file for ilab (optional)
+2. Initialize a taxonomy
+3. Ensure their taxonomy is valid
+4. Ingest some data for RAG and SDG (SDG coming soon)
+5. Generate synthetic data using an InstructLab pipeline
+6. Do some custom handling per their use case
+7. Fine-tune a model using the custom config they initialized for their hardware
+
+A user could do the following to accomplish this if they had an SDK:
+
+```python
+instructlab.core.configuration.init(taxonomy=path_to_taxonomy)
+if instructlab.core.taxonomy.diff():
+    instructlab.core.rag.ingest(...)
+    instructlab.core.data.ingest(...)
+    data_paths = instructlab.core.data.generate(...)
+    some_custom_handling(data_paths)
+    instructlab.core.model.train(data_path=..., strategy=lab_multiphase)
+```
```

Comment:
Not an issue with the proposal itself since the snippet is only illustrative, but I'd consider whether ilab SDK should promote using file paths for any artifacts. I'd rather have ilab encapsulate the knowledge about storage options (for models and datasets) and then allow SDK users to refer to the artifacts by their names. (This would be in line with the spirit of https://github.com/instructlab/instructlab/issues/1832 and some other issues under https://github.com/instructlab/instructlab/issues/1871)